### PR TITLE
Fixed failing tests on Edge

### LINKED ISSUE #1686: The '/tests/plugins/imagebase/features/link' tests are unstable on Edge
## Are you reporting a feature request or a bug?

Other: failing/unstable tests

## Provide detailed reproduction steps (if any)

1. Run `/tests/plugins/imagebase/features/link` test suite.

### Expected result

Whole test suit is passing.

### Actual result

Test suite fails randomly or is run only partially.

---

There are few different behaviours:

1. Tests passing, throwing some errors in the meantime:
screen shot 2018-02-16 at 15 56 25

2. After about 10 - 15 tests are run (all green), browser reloads window, runs tests again, reloads again (probably on the same test) and shows error.

3. Same as before, but there is one failing tests visible.

## Other details

* Browser: Edge (`16.16299`)
* OS: Windows 10
* CKEditor version: `4.9.0`
* Installed CKEditor plugins: `imagebase`

### PR DESCRIPTION
**Pull request by engineering-this**
## What is the purpose of this pull request?

Failing test fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [ ] Manual tests

## What changes did you make?

Most important was separating each test. Another thing was restoring selection in editor after selecting dialogs text input.

Closes #1686

### CODE REVIEW

File: tests/plugins/imagebase/features/link.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /* bender-tags: editor,widget */
 /* bender-ckeditor-plugins: imagebase,link,toolbar,contextmenu */
 /* bender-include: ../../widget/_helpers/tools.js */
+/* global widgetTestsTools */
```
Comment (engineering-this) [inline tests/plugins/imagebase/features/link.js]: Linters cried for it.

File: plugins/dialogui/plugin.js [hunk 2]
```diff
@@ -1011,14 +1011,22 @@ CKEDITOR.plugins.add( 'dialogui', {
...
 				setTimeout( function() {
-					var e = me.getInputElement();
+					var e = me.getInputElement(),
+						ranges = editor.getSelection().getRanges();
+
 					if ( e ) {
 						e.$.focus();
 						e.$.select();
+
+						// #1686 Edge occasionally loses editor selection. Restore it when it happens.
```
Comment (Comandeer) [inline plugins/dialogui/plugin.js]: It seems like a more general fix, which IMO should be moved to the separate issue and some tests should be added to it.

File: plugins/floatpanel/plugin.js [hunk 3]
```diff
@@ -366,19 +366,22 @@ CKEDITOR.plugins.add( 'floatpanel', {
...
-								// For IE8, there is not much logic on this, but it works.
-								if ( CKEDITOR.env.ie8Compat )
-									left -= element.getDocument().getDocumentElement().$.scrollLeft * 2;
-								else
-									left -= ( offsetParent.$.scrollWidth - offsetParent.$.clientWidth );
+							// #1686 preventing typeError within tests on Edge.
+							if ( scrollParent ) {
+								// Quirks returns <body>, but standards returns <html>.
+								if ( scrollParent.getName() == 'html' )
+									scrollParent = scrollParent.getDocument().getBody();
```
Comment (Comandeer) [inline plugins/floatpanel/plugin.js]: Please add missing curly braces.

File: plugins/floatpanel/plugin.js [hunk 4]
```diff
@@ -366,19 +366,22 @@ CKEDITOR.plugins.add( 'floatpanel', {
 						// position and horizontal scrolls. Here we have a
 						// series of hacks to workaround it. (https://dev.ckeditor.com/ticket/6146)
 						if ( CKEDITOR.env.ie ) {
-							var offsetParent = new CKEDITOR.dom.element( element.$.offsetParent ),
+							var offsetParent = element.$.offsetParent && new CKEDITOR.dom.element( element.$.offsetParent ),
```
[the remark this answers is not in the document]
Comment (engineering-this) [reply plugins/floatpanel/plugin.js]: Here is what I found:
>The HTMLElement.offsetParent (...) Returns null when the element has style.display set to "none".

So if that is correct behaviour not bug we should handle it.

[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/floatpanel/plugin.js]: OK, but the question is why IE wants to calculate anything for _hidden_ iframe? 😛

File: plugins/dialogui/plugin.js (hunk 2 above)
Comment (engineering-this) [reply plugins/dialogui/plugin.js]: I can't reproduce this problem manually. It occurs only in this case:

```
range.setStart( paragraph, 1 );
range.setEnd( paragraph, 3 );
range.select();
editor.execCommand( 'link' );
```

And focus is lost above in line 1023 by native method.

Comment (Comandeer) [reply plugins/dialogui/plugin.js]: Yet it's valid bug, connected with API. It should have dedicated ticket with tests etc.

Comment (engineering-this) [reply plugins/dialogui/plugin.js]: I spend some time testing it, and I can't reproduce this issue on Edge 17, hence I'm dropping that fix.

**Comment by engineering-this** [comment]
Rebased to current master and dropped edge fix for editor losing focus, as I can't reproduce it on Edge 17.

**Pull request merged**